// @stream-io/video-react-native-sdk/dist/commonjs/components/Livestream/LivestreamTopView/FollowerCount.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FollowerCount = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _contexts = require("../../../contexts");
var _videoReactBindings = require("@stream-io/video-react-bindings");
var _icons = require("../../../icons");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
/**
 * Props for the FollowerCount component.
 */

/**
 * The FollowerCount component that displays the number of participants while in the call.
 */
const FollowerCount = ({}) => {
  const styles = useStyles();
  const {
    theme: {
      followerCount
    }
  } = (0, _contexts.useTheme)();
  const {
    useParticipantCount
  } = (0, _videoReactBindings.useCallStateHooks)();
  const totalParticipants = useParticipantCount();
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, followerCount.container]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.icon, followerCount.icon]
  }, /*#__PURE__*/_react.default.createElement(_icons.Eye, null)), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: [styles.label, followerCount.label]
  }, totalParticipants));
};
exports.FollowerCount = FollowerCount;
const useStyles = () => {
  const {
    theme
  } = (0, _contexts.useTheme)();
  return (0, _react.useMemo)(() => _reactNative.StyleSheet.create({
    container: {
      paddingHorizontal: theme.variants.spacingSizes.sm,
      paddingVertical: 4,
      borderTopRightRadius: 4,
      borderBottomRightRadius: 4,
      flexDirection: 'row',
      alignItems: 'center',
      justifyContent: 'center',
      backgroundColor: theme.colors.sheetTertiary
    },
    icon: {
      height: theme.variants.iconSizes.sm,
      width: theme.variants.iconSizes.sm
    },
    label: {
      fontSize: theme.variants.fontSizes.md,
      fontWeight: '600',
      flexShrink: 1,
      textAlign: 'center',
      includeFontPadding: false,
      marginLeft: theme.variants.spacingSizes.xs,
      color: theme.colors.textPrimary
    }
  }), [theme]);
};
