// @stream-io/video-react-native-sdk/dist/commonjs/components/Livestream/LivestreamTopView/HostLivestreamTopView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HostLivestreamTopView = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _DurationBadge = require("./DurationBadge");
var _LiveIndicator = require("./LiveIndicator");
var _FollowerCount = require("./FollowerCount");
var _contexts = require("../../../contexts");
var _videoReactBindings = require("@stream-io/video-react-bindings");
var _constants = require("../../../constants");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * Props for the HostLivestreamTopView component.
 */

/**
 * The HostLivestreamTopView component displays the top view component of the host's live stream.
 */
const HostLivestreamTopView = ({
  DurationBadge = _DurationBadge.DurationBadge,
  LiveIndicator = _LiveIndicator.LiveIndicator,
  FollowerCount = _FollowerCount.FollowerCount,
  onLayout
}) => {
  const {
    useIsCallLive,
    useIsCallHLSBroadcastingInProgress
  } = (0, _videoReactBindings.useCallStateHooks)();
  const isCallLive = useIsCallLive();
  const isBroadcasting = useIsCallHLSBroadcastingInProgress();
  const liveOrBroadcasting = isCallLive || isBroadcasting;
  const {
    theme: {
      colors,
      hostLivestreamTopView
    }
  } = (0, _contexts.useTheme)();
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, {
      backgroundColor: colors.sheetOverlay
    }, hostLivestreamTopView.container],
    onLayout: onLayout
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.leftElement, hostLivestreamTopView.leftElement]
  }, DurationBadge && /*#__PURE__*/_react.default.createElement(DurationBadge, {
    mode: "host"
  })), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.centerElement, hostLivestreamTopView.centerElement]
  }), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.rightElement, hostLivestreamTopView.rightElement]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.liveInfo, hostLivestreamTopView.liveInfo]
  }, liveOrBroadcasting && LiveIndicator && /*#__PURE__*/_react.default.createElement(LiveIndicator, null), FollowerCount && /*#__PURE__*/_react.default.createElement(FollowerCount, null))));
};
exports.HostLivestreamTopView = HostLivestreamTopView;
const styles = _reactNative.StyleSheet.create({
  container: {
    position: 'absolute',
    top: 0,
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: 16,
    paddingHorizontal: 8,
    borderBottomEndRadius: 8,
    borderBottomStartRadius: 8,
    zIndex: _constants.Z_INDEX.IN_FRONT
  },
  liveInfo: {
    flexDirection: 'row'
  },
  leftElement: {
    flex: 1,
    alignItems: 'flex-start'
  },
  centerElement: {
    flex: 1,
    alignItems: 'center'
  },
  rightElement: {
    flex: 1,
    alignItems: 'flex-end'
  }
});
