// @stream-io/video-react-native-sdk/dist/commonjs/components/Livestream/LivestreamTopView/ViewerLivestreamTopView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ViewerLivestreamTopView = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _DurationBadge = require("./DurationBadge");
var _LiveIndicator = require("./LiveIndicator");
var _FollowerCount = require("./FollowerCount");
var _contexts = require("../../../contexts");
var _constants = require("../../../constants");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * Props for the ViewerLivestreamTopView component.
 */

/**
 * The ViewerLivestreamTopView component displays the top view component of the viewer's live stream.
 */
const ViewerLivestreamTopView = ({
  DurationBadge = _DurationBadge.DurationBadge,
  LiveIndicator = _LiveIndicator.LiveIndicator,
  FollowerCount = _FollowerCount.FollowerCount,
  onLayout
}) => {
  const {
    theme: {
      colors,
      viewerLivestreamTopView
    }
  } = (0, _contexts.useTheme)();
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, {
      backgroundColor: colors.sheetOverlay
    }, viewerLivestreamTopView.container],
    onLayout: onLayout
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.leftElement, viewerLivestreamTopView.leftElement]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.liveInfo, viewerLivestreamTopView.liveInfo]
  }, LiveIndicator && /*#__PURE__*/_react.default.createElement(LiveIndicator, null), FollowerCount && /*#__PURE__*/_react.default.createElement(FollowerCount, null))), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.centerElement, viewerLivestreamTopView.centerElement]
  }, DurationBadge && /*#__PURE__*/_react.default.createElement(DurationBadge, {
    mode: "viewer"
  })), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.rightElement, viewerLivestreamTopView.rightElement]
  }));
};
exports.ViewerLivestreamTopView = ViewerLivestreamTopView;
const styles = _reactNative.StyleSheet.create({
  container: {
    position: 'absolute',
    top: 0,
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: 16,
    paddingHorizontal: 8,
    borderBottomEndRadius: 8,
    borderBottomStartRadius: 8,
    zIndex: _constants.Z_INDEX.IN_FRONT
  },
  liveInfo: {
    flexDirection: 'row'
  },
  leftElement: {
    flex: 1,
    alignItems: 'flex-start'
  },
  centerElement: {
    flex: 1,
    alignItems: 'center',
    flexGrow: 3
  },
  rightElement: {
    flex: 1,
    alignItems: 'flex-end'
  }
});
