// @stream-io/video-react-native-sdk/dist/module/constants/TestIds.js
export let IconTestIds = /*#__PURE__*/function (IconTestIds) {
  IconTestIds["MUTED_VIDEO"] = "muted-video-icon";
  IconTestIds["HANG_UP_CALL"] = "hang-up-call-icon";
  IconTestIds["PHONE"] = "phone-icon";
  IconTestIds["SCREEN_SHARE_INDICATOR"] = "screen-share-indicator-icon";
  IconTestIds["SCREEN_SHARE"] = "screen-share-icon";
  return IconTestIds;
}({});
export let ComponentTestIds = /*#__PURE__*/function (ComponentTestIds) {
  ComponentTestIds["PARTICIPANT_AVATAR"] = "participant-avatar";
  ComponentTestIds["CALL_PARTICIPANTS_LIST"] = "call-participants-list";
  ComponentTestIds["CALL_PARTICIPANTS_SPOTLIGHT"] = "call-participants-spotlight";
  ComponentTestIds["CALL_PARTICIPANTS_GRID"] = "call-participants-grid";
  ComponentTestIds["LOCAL_PARTICIPANT"] = "local-participant";
  ComponentTestIds["PARTICIPANT_MEDIA_STREAM"] = "participant-media-stream";
  ComponentTestIds["PARTICIPANTS_INFO"] = "participants-info";
  ComponentTestIds["PARTICIPANT_SCREEN_SHARING"] = "participant-screen-sharing";
  ComponentTestIds["REACTIONS_PICKER"] = "reactions-picker";
  return ComponentTestIds;
}({});
export let ButtonTestIds = /*#__PURE__*/function (ButtonTestIds) {
  ButtonTestIds["PARTICIPANTS_INFO"] = "participants-info-button";
  ButtonTestIds["PARTICIPANTS_COUNT_VIEW"] = "participants-count-view";
  ButtonTestIds["PARTICIPANTS_COUNT_TEXT"] = "participants-count-text";
  ButtonTestIds["HANG_UP_CALL"] = "hang-up-call";
  ButtonTestIds["REACTION"] = "call-controls-reaction";
  return ButtonTestIds;
}({});
export let ImageTestIds = /*#__PURE__*/function (ImageTestIds) {
  ImageTestIds["AVATAR"] = "avatar-image";
  return ImageTestIds;
}({});
