// @stream-io/video-react-native-sdk/dist/module/components/Livestream/LivestreamControls/HostLivestreamControls.js
import React from 'react';
import { StyleSheet, View } from 'react-native';
import { HostStartStreamButton as DefaultHostStartStreamButton } from './HostStartStreamButton';
import { LivestreamMediaControls as DefaultLivestreamMediaControls } from './LivestreamMediaControls';
import { useTheme } from '../../../contexts';
import { Z_INDEX } from '../../../constants';

/**
 * Props for the HostLivestreamControls component.
 */

/**
 * The HostLivestreamControls component displays the call controls for the live stream at host's end.
 */
export const HostLivestreamControls = ({
  HostStartStreamButton = DefaultHostStartStreamButton,
  LivestreamMediaControls = DefaultLivestreamMediaControls,
  onEndStreamHandler,
  onStartStreamHandler,
  hls,
  disableStopPublishedStreamsOnEndStream,
  onLayout
}) => {
  const {
    theme: {
      colors,
      hostLivestreamControls
    }
  } = useTheme();
  return /*#__PURE__*/React.createElement(View, {
    style: [styles.container, {
      backgroundColor: colors.sheetOverlay
    }, hostLivestreamControls.container],
    onLayout: onLayout
  }, /*#__PURE__*/React.createElement(View, {
    style: [styles.leftElement, hostLivestreamControls.leftElement]
  }, HostStartStreamButton && /*#__PURE__*/React.createElement(HostStartStreamButton, {
    onEndStreamHandler: onEndStreamHandler,
    onStartStreamHandler: onStartStreamHandler,
    hls: hls,
    disableStopPublishedStreamsOnEndStream: disableStopPublishedStreamsOnEndStream
  })), /*#__PURE__*/React.createElement(View, {
    style: [styles.rightElement, hostLivestreamControls.rightElement]
  }, LivestreamMediaControls && /*#__PURE__*/React.createElement(LivestreamMediaControls, null)));
};
const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    bottom: 0,
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: 16,
    paddingHorizontal: 8,
    zIndex: Z_INDEX.IN_FRONT
  },
  leftElement: {
    flex: 1,
    alignItems: 'flex-start'
  },
  rightElement: {
    flex: 1,
    alignItems: 'flex-end'
  }
});
