// @stream-io/video-react-native-sdk/dist/commonjs/components/Livestream/LivestreamControls/HostLivestreamControls.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HostLivestreamControls = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _HostStartStreamButton = require("./HostStartStreamButton");
var _LivestreamMediaControls = require("./LivestreamMediaControls");
var _contexts = require("../../../contexts");
var _constants = require("../../../constants");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * Props for the HostLivestreamControls component.
 */

/**
 * The HostLivestreamControls component displays the call controls for the live stream at host's end.
 */
const HostLivestreamControls = ({
  HostStartStreamButton = _HostStartStreamButton.HostStartStreamButton,
  LivestreamMediaControls = _LivestreamMediaControls.LivestreamMediaControls,
  onEndStreamHandler,
  onStartStreamHandler,
  hls,
  disableStopPublishedStreamsOnEndStream,
  onLayout
}) => {
  const {
    theme: {
      colors,
      hostLivestreamControls
    }
  } = (0, _contexts.useTheme)();
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, {
      backgroundColor: colors.sheetOverlay
    }, hostLivestreamControls.container],
    onLayout: onLayout
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.leftElement, hostLivestreamControls.leftElement]
  }, HostStartStreamButton && /*#__PURE__*/_react.default.createElement(HostStartStreamButton, {
    onEndStreamHandler: onEndStreamHandler,
    onStartStreamHandler: onStartStreamHandler,
    hls: hls,
    disableStopPublishedStreamsOnEndStream: disableStopPublishedStreamsOnEndStream
  })), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.rightElement, hostLivestreamControls.rightElement]
  }, LivestreamMediaControls && /*#__PURE__*/_react.default.createElement(LivestreamMediaControls, null)));
};
exports.HostLivestreamControls = HostLivestreamControls;
const styles = _reactNative.StyleSheet.create({
  container: {
    position: 'absolute',
    bottom: 0,
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: 16,
    paddingHorizontal: 8,
    zIndex: _constants.Z_INDEX.IN_FRONT
  },
  leftElement: {
    flex: 1,
    alignItems: 'flex-start'
  },
  rightElement: {
    flex: 1,
    alignItems: 'flex-end'
  }
});
