// @stream-io/video-react-native-sdk/dist/commonjs/components/Livestream/LivestreamControls/LivestreamAudioControlButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LivestreamAudioControlButton = void 0;
var _videoReactBindings = require("@stream-io/video-react-bindings");
var _react = _interopRequireDefault(require("react"));
var _contexts = require("../../../contexts");
var _reactNative = require("react-native");
var _icons = require("../../../icons");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * The LivestreamAudioControlButton controls the audio stream publish/unpublish while in the livestream for the host.
 */
const LivestreamAudioControlButton = () => {
  const {
    useMicrophoneState
  } = (0, _videoReactBindings.useCallStateHooks)();
  const {
    optimisticIsMute,
    microphone
  } = useMicrophoneState();
  const {
    theme: {
      colors,
      variants: {
        iconSizes,
        buttonSizes
      },
      livestreamAudioControlButton
    }
  } = (0, _contexts.useTheme)();
  const onPress = async () => {
    await microphone.toggle();
  };
  return /*#__PURE__*/_react.default.createElement(_reactNative.Pressable, {
    onPress: onPress,
    style: [styles.container, {
      backgroundColor: colors.buttonSecondary,
      height: buttonSizes.xs,
      width: buttonSizes.xs
    }, livestreamAudioControlButton.container]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.icon, {
      height: iconSizes.sm,
      width: iconSizes.sm
    }, livestreamAudioControlButton.icon]
  }, /*#__PURE__*/_react.default.createElement(_icons.IconWrapper, null, !optimisticIsMute ? /*#__PURE__*/_react.default.createElement(_icons.Mic, {
    color: colors.iconPrimary,
    size: iconSizes.md
  }) : /*#__PURE__*/_react.default.createElement(_icons.MicOff, {
    color: colors.iconPrimary,
    size: iconSizes.md
  }))));
};
exports.LivestreamAudioControlButton = LivestreamAudioControlButton;
const styles = _reactNative.StyleSheet.create({
  container: {
    justifyContent: 'center',
    alignItems: 'center',
    marginHorizontal: 4,
    borderRadius: 4
  },
  icon: {}
});
