// @stream-io/video-react-native-sdk/dist/commonjs/components/Livestream/LivestreamControls/LivestreamMediaControls.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LivestreamMediaControls = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _contexts = require("../../../contexts");
var _LivestreamAudioControlButton = require("./LivestreamAudioControlButton");
var _LivestreamVideoControlButton = require("./LivestreamVideoControlButton");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * Props for the LivestreamMediaControls component.
 */

/**
 * The LivestreamMediaControls component controls the media publish/unpublish for the host's live stream.
 */
const LivestreamMediaControls = ({}) => {
  const {
    theme: {
      livestreamMediaControls
    }
  } = (0, _contexts.useTheme)();
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, livestreamMediaControls.container]
  }, /*#__PURE__*/_react.default.createElement(_LivestreamAudioControlButton.LivestreamAudioControlButton, null), /*#__PURE__*/_react.default.createElement(_LivestreamVideoControlButton.LivestreamVideoControlButton, null));
};
exports.LivestreamMediaControls = LivestreamMediaControls;
const styles = _reactNative.StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center'
  }
});
