// @stream-io/video-react-native-sdk/dist/module/theme/theme.js
import { colors } from './colors';
export const defaultTheme = {
  variants: {
    roundButtonSizes: {
      xs: 16,
      sm: 24,
      md: 36,
      lg: 44,
      xl: 56
    },
    borderRadiusSizes: {
      xs: 4,
      sm: 8,
      md: 16,
      lg: 24,
      xl: 32
    },
    buttonSizes: {
      xs: 40,
      sm: 50,
      md: 60,
      lg: 70,
      xl: 80
    },
    iconSizes: {
      xs: 15,
      sm: 20,
      md: 25,
      lg: 30,
      xl: 35
    },
    avatarSizes: {
      xs: 50,
      sm: 90,
      md: 100,
      lg: 160,
      xl: 180
    },
    spacingSizes: {
      xs: 4,
      sm: 8,
      md: 16,
      lg: 24,
      xl: 32
    },
    fontSizes: {
      xs: 8,
      sm: 12,
      md: 16,
      lg: 20,
      xl: 24
    },
    insets: {
      top: 0,
      right: 0,
      bottom: 0,
      left: 0
    }
  },
  typefaces: {
    heading4: {
      fontSize: 34,
      fontWeight: '400'
    },
    heading5: {
      fontSize: 24,
      fontWeight: '400'
    },
    heading6: {
      fontSize: 20,
      fontWeight: '500'
    },
    subtitle: {
      fontSize: 16,
      fontWeight: '400'
    },
    subtitleBold: {
      fontSize: 16,
      fontWeight: '500'
    },
    bodyBold: {
      fontSize: 16,
      fontWeight: '600'
    },
    caption: {
      fontSize: 10,
      fontWeight: '400'
    }
  },
  defaults: {
    color: colors.primary,
    backgroundColor: colors.sheetPrimary,
    margin: 10,
    padding: 10,
    fontSize: 16,
    fontWeight: '500',
    borderRadius: 32,
    iconSize: 28,
    borderColor: colors.buttonPrimary,
    borderWidth: 1
  },
  colors: colors,
  avatar: {
    container: {},
    image: {},
    text: {}
  },
  acceptCallButton: {
    container: {},
    svgContainer: {}
  },
  callContent: {
    container: {},
    callParticipantsContainer: {},
    topContainer: {}
  },
  callControls: {
    container: {}
  },
  callControlsButton: {
    container: {},
    svgContainer: {}
  },
  joinCallButton: {
    container: {},
    label: {}
  },
  chatButton: {
    container: {},
    svgContainer: {}
  },
  hangupCallButton: {
    container: {},
    svgContainer: {}
  },
  rejectCallButton: {
    container: {},
    svgContainer: {}
  },
  screenShareToggleButton: {
    container: {},
    svgContainer: {}
  },
  toggleAudioPreviewButton: {
    container: {},
    svgContainer: {}
  },
  toggleAudioPublishingButton: {
    container: {},
    svgContainer: {}
  },
  toggleCameraFaceButton: {
    container: {},
    svgContainer: {}
  },
  toggleVideoPreviewButton: {
    container: {},
    svgContainer: {}
  },
  toggleVideoPublishingButton: {
    container: {},
    svgContainer: {}
  },
  callParticipantsGrid: {
    container: {}
  },
  callParticipantsSpotlight: {
    container: {},
    fullScreenSpotlightContainer: {},
    callParticipantsListContainer: {},
    spotlightContainer: {}
  },
  lobby: {
    container: {},
    topContainer: {},
    heading: {},
    subHeading: {},
    bottomContainer: {},
    videoContainer: {},
    infoContainer: {},
    infoText: {},
    participantStatusContainer: {},
    avatarContainer: {},
    userNameLabel: {},
    audioMutedIconContainer: {}
  },
  floatingParticipantsView: {
    container: {},
    participantViewContainer: {},
    videoFallback: {}
  },
  participantLabel: {
    container: {},
    userNameLabel: {},
    audioMutedIconContainer: {},
    videoMutedIconContainer: {},
    pinIconContainer: {},
    screenShareIconContainer: {}
  },
  participantNetworkQualityIndicator: {
    container: {}
  },
  participantReaction: {
    container: {},
    reaction: {}
  },
  participantVideoFallback: {
    container: {},
    label: {},
    avatarContainer: {},
    avatarText: {},
    avatarImage: {}
  },
  participantView: {
    container: {},
    footerContainer: {},
    highlightedContainer: {}
  },
  reactionsPicker: {
    reactionsPopup: {},
    reactionsButtonDimmer: {},
    reactionItem: {},
    reactionText: {}
  },
  videoRenderer: {
    container: {},
    videoStream: {}
  },
  incomingCall: {
    background: {},
    content: {},
    topContainer: {},
    incomingCallText: {},
    incomingCallControls: {},
    bottomContainer: {},
    buttonGroup: {}
  },
  outgoingCall: {
    container: {},
    background: {},
    content: {},
    topContainer: {},
    callingText: {},
    outgoingCallControls: {},
    bottomContainer: {},
    buttonGroup: {},
    deviceControlButtons: {}
  },
  ringingCallContent: {
    container: {}
  },
  userInfo: {
    container: {},
    avatarGroup: {},
    name: {}
  },
  lobbyControls: {
    container: {}
  },
  participantInfoBadge: {
    container: {},
    participantCountContainer: {},
    participantsIconContainer: {},
    participantsCountText: {}
  },
  hostLivestream: {
    container: {}
  },
  hostLivestreamTopView: {
    container: {},
    leftElement: {},
    rightElement: {},
    centerElement: {},
    liveInfo: {}
  },
  viewerLivestream: {
    container: {}
  },
  livestreamLayout: {
    container: {}
  },
  viewerLivestreamTopView: {
    container: {},
    leftElement: {},
    centerElement: {},
    rightElement: {},
    liveInfo: {}
  },
  followerCount: {
    container: {},
    icon: {},
    label: {}
  },
  durationBadge: {
    container: {},
    icon: {},
    label: {}
  },
  liveIndicator: {
    container: {},
    label: {}
  },
  hostLivestreamControls: {
    container: {},
    leftElement: {},
    rightElement: {}
  },
  viewerLivestreamControls: {
    container: {},
    leftElement: {},
    rightElement: {}
  },
  hostStartStreamButton: {
    container: {},
    icon: {},
    text: {}
  },
  viewerLeaveStreamButton: {
    container: {},
    icon: {},
    text: {}
  },
  livestreamMediaControls: {
    container: {}
  },
  livestreamAudioControlButton: {
    container: {},
    icon: {}
  },
  livestreamVideoControlButton: {
    container: {},
    icon: {}
  },
  livestreamScreenShareToggleButton: {
    container: {},
    icon: {}
  },
  screenshareOverlay: {
    container: {},
    text: {},
    button: {},
    buttonIcon: {},
    buttonText: {}
  }
};
