// @stream-io/video-react-native-sdk/dist/commonjs/components/utility/Avatar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Avatar = void 0;
var _reactNative = require("react-native");
var _react = _interopRequireDefault(require("react"));
var _utils = require("../../utils");
var _TestIds = require("../../constants/TestIds");
var _ThemeContext = require("../../contexts/ThemeContext");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * Props to be passed for the Avatar component.
 */

/**
 * Shows either user's image or initials based on the user state and existence of
 * their image.
 */
const Avatar = props => {
  const {
    participant: {
      userId,
      image,
      name
    },
    size = 100,
    style: styleProp
  } = props;
  const {
    theme: {
      avatar,
      colors,
      typefaces
    }
  } = (0, _ThemeContext.useTheme)();
  const userDetails = name || userId;
  const userLabel = userDetails ? (0, _utils.getInitialsOfName)(userDetails) : '?';
  const imageUrl = image;
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    testID: _TestIds.ComponentTestIds.PARTICIPANT_AVATAR,
    style: [styles.container, {
      borderRadius: size / 2,
      height: size,
      width: size
    }, {
      backgroundColor: colors.primary
    }, avatar.container, styleProp?.container]
  }, imageUrl ? /*#__PURE__*/_react.default.createElement(_reactNative.Image, {
    testID: _TestIds.ImageTestIds.AVATAR,
    source: {
      uri: imageUrl
    },
    style: [styles.image, avatar.image, styleProp?.image]
  }) : /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: [styles.text, {
      fontSize: size / 2,
      color: colors.textPrimary
    }, typefaces.heading6, avatar.text, styleProp?.text],
    numberOfLines: 1
  }, userLabel));
};
exports.Avatar = Avatar;
const styles = _reactNative.StyleSheet.create({
  container: {
    justifyContent: 'center',
    overflow: 'hidden'
  },
  image: {
    flex: 1
  },
  text: {
    textAlign: 'center'
  }
});
