// @stream-io/video-react-native-sdk/dist/commonjs/components/Call/Lobby/LobbyFooter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LobbyFooter = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _ThemeContext = require("../../../contexts/ThemeContext");
var _videoReactBindings = require("@stream-io/video-react-bindings");
var _Lock = require("../../../icons/Lock");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
/**
 * Props for the Lobby Footer in the Lobby component.
 */

/**
 * The default Lobby Footer to be used in the Lobby component.
 */
const LobbyFooter = ({
  onJoinCallHandler,
  JoinCallButton
}) => {
  const {
    theme: {
      colors,
      lobby,
      variants
    }
  } = (0, _ThemeContext.useTheme)();
  const styles = useStyles();
  const {
    useCallSession
  } = (0, _videoReactBindings.useCallStateHooks)();
  const {
    t
  } = (0, _videoReactBindings.useI18n)();
  const session = useCallSession();
  const numberOfParticipants = session?.participants.length;
  const participantsText = (0, _react.useMemo)(() => {
    if (!numberOfParticipants) {
      return t('Currently there are no other participants in the call.');
    }
    if (numberOfParticipants === 1) {
      return t('There is {{numberOfParticipants}} more person in the call.', {
        numberOfParticipants
      });
    }
    return t('There are {{numberOfParticipants}} more people in the call.', {
      numberOfParticipants
    });
  }, [numberOfParticipants, t]);
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.mainContainer, lobby.infoContainer]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.textContainer
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.iconContainer
  }, /*#__PURE__*/_react.default.createElement(_Lock.Lock, {
    color: colors.iconPrimary,
    size: variants.iconSizes.md
  })), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: [{
      color: colors.textPrimary
    }, styles.infoText, lobby.infoText]
  }, t('You are about to join a call.') + ' ' + participantsText)), JoinCallButton && /*#__PURE__*/_react.default.createElement(JoinCallButton, {
    onJoinCallHandler: onJoinCallHandler
  }));
};
exports.LobbyFooter = LobbyFooter;
const useStyles = () => {
  const {
    theme
  } = (0, _ThemeContext.useTheme)();
  return (0, _react.useMemo)(() => _reactNative.StyleSheet.create({
    mainContainer: {
      padding: theme.variants.spacingSizes.sm
    },
    textContainer: {
      flexDirection: 'row',
      alignItems: 'center',
      justifyContent: 'center',
      backgroundColor: theme.colors.sheetTertiary,
      paddingHorizontal: theme.variants.spacingSizes.md,
      paddingVertical: theme.variants.spacingSizes.xs,
      borderRadius: theme.variants.borderRadiusSizes.sm
    },
    iconContainer: {
      marginRight: theme.variants.spacingSizes.sm
    },
    infoText: {
      fontSize: theme.variants.fontSizes.sm,
      lineHeight: 20,
      fontWeight: '400'
    }
  }), [theme]);
};
