// @stream-io/video-react-native-sdk/dist/commonjs/components/Call/Lobby/Lobby.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Lobby = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _videoReactBindings = require("@stream-io/video-react-bindings");
var _Avatar = require("../../utility/Avatar");
var _reactNativeWebrtc = require("@stream-io/react-native-webrtc");
var _LobbyControls = require("../CallControls/LobbyControls");
var _JoinCallButton = require("./JoinCallButton");
var _ThemeContext = require("../../../contexts/ThemeContext");
var _LobbyFooter = require("./LobbyFooter");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
/**
 * Props for the Lobby Component.
 */

/**
 * Components that acts as a pre-join view for the call. Where you can preview your video and audio. Check for call details and check for number of participants already in the call.
 */
const Lobby = ({
  onJoinCallHandler,
  LobbyControls = _LobbyControls.LobbyControls,
  JoinCallButton = _JoinCallButton.JoinCallButton,
  landscape = false,
  LobbyFooter = _LobbyFooter.LobbyFooter
}) => {
  const {
    theme: {
      colors,
      lobby,
      typefaces
    }
  } = (0, _ThemeContext.useTheme)();
  const styles = useStyles(landscape);
  const connectedUser = (0, _videoReactBindings.useConnectedUser)();
  const {
    useCameraState,
    useCallSettings
  } = (0, _videoReactBindings.useCallStateHooks)();
  const callSettings = useCallSettings();
  const isVideoEnabledInCall = callSettings?.video.enabled;
  const {
    isMute: cameraIsMuted,
    mediaStream
  } = useCameraState();
  const {
    t
  } = (0, _videoReactBindings.useI18n)();
  const localVideoStream = mediaStream;
  const connectedUserAsParticipant = {
    userId: connectedUser?.id,
    image: connectedUser?.image,
    name: connectedUser?.name
  };
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, lobby.container]
  }, connectedUser && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: [styles.heading, typefaces.heading5, lobby.heading]
  }, t('Before joining')), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: [styles.subHeading, lobby.subHeading]
  }, t('Setup your audio and video')), isVideoEnabledInCall && /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.videoContainer, {
      backgroundColor: colors.sheetTertiary
    }, lobby.videoContainer]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.topView
  }), !cameraIsMuted && localVideoStream ? /*#__PURE__*/_react.default.createElement(_reactNativeWebrtc.RTCView, {
    mirror: true,
    streamURL: localVideoStream.toURL(),
    objectFit: "cover",
    style: _reactNative.StyleSheet.absoluteFillObject
  }) : /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.avatarContainer, lobby.avatarContainer]
  }, /*#__PURE__*/_react.default.createElement(_Avatar.Avatar, {
    participant: connectedUserAsParticipant
  })), /*#__PURE__*/_react.default.createElement(ParticipantStatus, null))), LobbyControls && /*#__PURE__*/_react.default.createElement(LobbyControls, null), LobbyFooter && /*#__PURE__*/_react.default.createElement(LobbyFooter, {
    JoinCallButton: JoinCallButton,
    onJoinCallHandler: onJoinCallHandler
  }));
};
exports.Lobby = Lobby;
const ParticipantStatus = () => {
  const {
    theme: {
      colors,
      typefaces,
      lobby
    }
  } = (0, _ThemeContext.useTheme)();
  const styles = useStyles();
  const connectedUser = (0, _videoReactBindings.useConnectedUser)();
  const participantLabel = connectedUser?.name ?? connectedUser?.id;
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.participantStatusContainer, {
      backgroundColor: colors.sheetOverlay
    }, lobby.participantStatusContainer]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: [styles.userNameLabel, {
      color: colors.textPrimary
    }, typefaces.caption, lobby.userNameLabel],
    numberOfLines: 1
  }, participantLabel));
};
const useStyles = (landscape = false) => {
  const {
    theme
  } = (0, _ThemeContext.useTheme)();
  return (0, _react.useMemo)(() => _reactNative.StyleSheet.create({
    heading: {
      textAlign: 'center',
      color: theme.colors.textPrimary,
      paddingBottom: theme.variants.spacingSizes.xs
    },
    subHeading: {
      textAlign: 'center',
      paddingBottom: theme.variants.spacingSizes.md,
      color: theme.colors.textSecondary
    },
    container: {
      flex: 1,
      justifyContent: 'center',
      backgroundColor: theme.colors.sheetPrimary,
      paddingRight: theme.variants.insets.right + theme.variants.spacingSizes.sm,
      paddingLeft: theme.variants.insets.left + theme.variants.spacingSizes.sm,
      paddingTop: theme.variants.insets.top,
      paddingBottom: theme.variants.insets.bottom
    },
    videoContainer: {
      height: landscape ? '40%' : '50%',
      borderRadius: theme.variants.borderRadiusSizes.md,
      justifyContent: 'space-between',
      alignItems: 'center',
      overflow: 'hidden'
    },
    topView: {},
    participantStatusContainer: {
      alignSelf: 'flex-start',
      flexDirection: 'row',
      alignItems: 'center',
      padding: theme.variants.spacingSizes.sm,
      borderTopRightRadius: theme.variants.borderRadiusSizes.sm
    },
    avatarContainer: {
      flex: 2,
      justifyContent: 'center'
    },
    userNameLabel: {
      flexShrink: 1
    }
  }), [theme, landscape]);
};
