// @stream-io/video-react-native-sdk/dist/module/components/Participant/ParticipantView/ParticipantVideoFallback.js
import React from 'react';
import { StyleSheet, Text, View } from 'react-native';
import { Avatar } from '../../utility';
import { useTheme } from '../../../contexts/ThemeContext';

/**
 * Props for the ParticipantVideoFallback component.
 */

/**
 * This component is used to customize the video fallback of the participant, when the video is disabled.
 */
export const ParticipantVideoFallback = ({
  participant
}) => {
  const {
    theme: {
      colors,
      typefaces,
      participantVideoFallback
    }
  } = useTheme();
  const {
    name,
    image,
    userId
  } = participant;
  const participantLabel = name ?? userId;

  // Display the Participant name/user id if the image isn't present.
  return /*#__PURE__*/React.createElement(View, {
    style: [styles.container, {
      backgroundColor: colors.sheetTertiary
    }, participantVideoFallback.container]
  }, !image ? /*#__PURE__*/React.createElement(Text, {
    style: [{
      color: colors.textPrimary
    }, typefaces.bodyBold, participantVideoFallback.label]
  }, participantLabel) : /*#__PURE__*/React.createElement(Avatar, {
    participant: participant,
    style: {
      container: participantVideoFallback.avatarContainer,
      image: participantVideoFallback.avatarImage,
      text: participantVideoFallback.avatarText
    }
  }));
};
const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    justifyContent: 'center',
    ...StyleSheet.absoluteFillObject
  }
});
