// @stream-io/video-react-native-sdk/dist/module/components/Participant/ParticipantView/ParticipantNetworkQualityIndicator.js
import { StyleSheet, View } from 'react-native';
import { Path, Svg } from 'react-native-svg';
import { SfuModels } from '@stream-io/video-client';
import React from 'react';
import { Z_INDEX } from '../../../constants';
import { useTheme } from '../../../contexts/ThemeContext';

/**
 * Props for the NetworkQualityIndicator component.
 */

const useConnectionQualitySignalColors = participant => {
  const {
    theme: {
      colors
    }
  } = useTheme();
  const {
    connectionQuality
  } = participant;
  if (!connectionQuality) {
    return null;
  }
  switch (connectionQuality) {
    case SfuModels.ConnectionQuality.EXCELLENT:
      return [colors.iconSuccess, colors.iconSuccess, colors.iconSuccess];
    case SfuModels.ConnectionQuality.GOOD:
      return [colors.iconSuccess, colors.iconSuccess, colors.iconPrimary];
    case SfuModels.ConnectionQuality.POOR:
      return [colors.iconWarning, colors.iconPrimary, colors.iconPrimary];
    default:
      return null;
  }
};
export const ParticipantNetworkQualityIndicator = ({
  participant
}) => {
  const {
    theme: {
      colors,
      variants: {
        iconSizes
      },
      participantNetworkQualityIndicator
    }
  } = useTheme();
  const connectionQualityColors = useConnectionQualitySignalColors(participant);
  if (!connectionQualityColors) {
    return null;
  }
  return /*#__PURE__*/React.createElement(View, {
    style: [styles.container, {
      backgroundColor: colors.sheetOverlay,
      height: iconSizes.lg,
      width: iconSizes.lg
    }, participantNetworkQualityIndicator.container]
  }, /*#__PURE__*/React.createElement(Svg, {
    viewBox: "0 0 34 34",
    fill: 'none'
  }, /*#__PURE__*/React.createElement(Path, {
    d: "M 9.97559 22.3379 L 9.97559 19.616",
    stroke: connectionQualityColors[0],
    strokeWidth: 3,
    strokeLinecap: "round",
    strokeLinejoin: "round",
    fill: connectionQualityColors[0]
  }), /*#__PURE__*/React.createElement(Path, {
    d: "M 16.7808 22.3379 L 16.7808 15.5331",
    stroke: connectionQualityColors[1],
    strokeWidth: 3,
    strokeLinecap: "round",
    strokeLinejoin: "round",
    fill: connectionQualityColors[1]
  }), /*#__PURE__*/React.createElement(Path, {
    d: "M 23.5854 22.3379 L 23.5854 11.4502",
    stroke: connectionQualityColors[2],
    strokeWidth: 3,
    strokeLinecap: "round",
    strokeLinejoin: "round",
    fill: connectionQualityColors[2]
  })));
};
const styles = StyleSheet.create({
  container: {
    zIndex: Z_INDEX.IN_FRONT,
    alignSelf: 'flex-end',
    borderTopLeftRadius: 5
  }
});
