// @stream-io/video-react-native-sdk/dist/commonjs/utils/push/internal/rxSubjects.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.voipPushNotificationCallCId$ = exports.voipCallkeepCallOnForegroundMap$ = exports.voipCallkeepAcceptedCallOnNativeDialerMap$ = exports.pushUnsubscriptionCallbacks$ = exports.pushTappedIncomingCallCId$ = exports.pushRejectedIncomingCallCId$ = exports.pushNonRingingCallData$ = exports.pushAndroidBackgroundDeliveredIncomingCallCId$ = exports.pushAcceptedIncomingCallCId$ = void 0;
var _rxjs = require("rxjs");
/**
 * This rxjs subject is used to store the call cid of the accepted incoming call from push notification
 * Note: it is should be subscribed only when a user has connected to the websocket of Stream
 */
const pushNonRingingCallData$ = exports.pushNonRingingCallData$ = new _rxjs.BehaviorSubject(undefined);

/**
 * This rxjs subject is used to store the call cid of the accepted incoming call from push notification
 * Note: it is should be subscribed only when a user has connected to the websocket of Stream
 */
const pushAcceptedIncomingCallCId$ = exports.pushAcceptedIncomingCallCId$ = new _rxjs.BehaviorSubject(undefined);

/**
 * This rxjs subject is used to store the call cid of the tapped incoming call from push notification it is neither accepted nor rejected yet
 * Note: it should be subscribed only when a user has connected to the websocket of Stream
 */
const pushTappedIncomingCallCId$ = exports.pushTappedIncomingCallCId$ = new _rxjs.BehaviorSubject(undefined);

/**
 * This rxjs subject is used to store the call cid of the delivered incoming call from push notification it is neither accepted nor rejected yet
 * Used so that the call is navigated to when app is open from being killed
 * Note: it should be subscribed only when a user has connected to the websocket of Stream
 */
const pushAndroidBackgroundDeliveredIncomingCallCId$ = exports.pushAndroidBackgroundDeliveredIncomingCallCId$ = new _rxjs.BehaviorSubject(undefined);

/**
 * This rxjs subject is used to store the call cid of the accepted incoming call from push notification
 * Note: it should be subscribed only when a user has connected to the websocket of Stream
 */
const pushRejectedIncomingCallCId$ = exports.pushRejectedIncomingCallCId$ = new _rxjs.BehaviorSubject(undefined);

/**
 * This rxjs subject is used to store the call cid of the incoming call from ios voip pushkit notification
 */
const voipPushNotificationCallCId$ = exports.voipPushNotificationCallCId$ = new _rxjs.BehaviorSubject(undefined);

/** The pair of cid of a call and its corresponding uuid created in the native side */

/*
 * This rxjs subject should only used to store the CallkeepMap
 * for the incoming call when on foreground
 * or in other words, when we get didDisplayIncomingCall from callkeep lib
 */
const voipCallkeepCallOnForegroundMap$ = exports.voipCallkeepCallOnForegroundMap$ = new _rxjs.BehaviorSubject(undefined);

/*
 * This rxjs subject should only used to store the CallkeepMap when it was accepted in the native dialer
 */
const voipCallkeepAcceptedCallOnNativeDialerMap$ = exports.voipCallkeepAcceptedCallOnNativeDialerMap$ = new _rxjs.BehaviorSubject(undefined);
/**
 * This rxjs subject is used to store the unsubscribe callbacks (if any) of the push notification processing
 * Note: it should be used to clear it when app processes push notification from foreground
 */
const pushUnsubscriptionCallbacks$ = exports.pushUnsubscriptionCallbacks$ = new _rxjs.BehaviorSubject(undefined);
