// @stream-io/video-react-native-sdk/dist/commonjs/providers/StreamVideo.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StreamVideo = void 0;
var _videoReactBindings = require("@stream-io/video-react-bindings");
var _react = _interopRequireWildcard(require("react"));
var _netinfo = _interopRequireDefault(require("@react-native-community/netinfo"));
var _hooks = require("../hooks");
var _translations = require("../translations");
var _ThemeContext = require("../contexts/ThemeContext");
var _ScreenshotIosContext = require("../contexts/internal/ScreenshotIosContext");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
/**
 *
 * @param props
 * @returns
 *
 * @category Client State
 */
const StreamVideo = props => {
  const {
    client,
    children,
    translationsOverrides = _translations.translations,
    i18nInstance,
    language,
    style
  } = props;

  /**
   * Effect to inform the coordinator about the online status of the app
   */
  (0, _react.useEffect)(() => {
    let prevIsOnline = true;
    return _netinfo.default.addEventListener(state => {
      const {
        isConnected,
        isInternetReachable
      } = state;
      const isOnline = isConnected === true && isInternetReachable !== false;
      if (isOnline === prevIsOnline) {
        return;
      }
      prevIsOnline = isOnline;
      const type = isOnline ? 'online' : 'offline';
      client.streamClient.updateNetworkConnectionStatus({
        type
      });
    });
  }, [client]);
  return /*#__PURE__*/_react.default.createElement(_videoReactBindings.StreamVideoProvider, {
    client: client,
    language: language,
    translationsOverrides: translationsOverrides,
    i18nInstance: i18nInstance
  }, /*#__PURE__*/_react.default.createElement(_ThemeContext.StreamTheme, {
    style: style
  }, /*#__PURE__*/_react.default.createElement(_ScreenshotIosContext.ScreenshotIosContextProvider, null, /*#__PURE__*/_react.default.createElement(PushRegister, null), children)));
};

/**
 * The usePushRegisterEffect needs to be a child of StreamVideoStoreProvider
 * So we create a renderless component to use it
 */
exports.StreamVideo = StreamVideo;
const PushRegister = () => {
  (0, _hooks.usePushRegisterEffect)();
  return null;
};
