// @stream-io/video-react-native-sdk/dist/module/components/Call/RingingCallContent/IncomingCall.js
import React from 'react';
import { ImageBackground, StyleSheet, Text, View } from 'react-native';
import { useCallStateHooks, useConnectedUser, useI18n } from '@stream-io/video-react-bindings';
import { UserInfo } from './UserInfo';
import { IncomingCallControls as DefaultIncomingCallControls } from '../CallControls';
import { useTheme } from '../../../contexts';

/**
 * Props for the IncomingCall Component.
 */

/**
 * An incoming call with the caller's avatar, name and accept/reject buttons.
 * Used when the user is receiving a call.
 */
export const IncomingCall = ({
  onAcceptCallHandler,
  onRejectCallHandler,
  IncomingCallControls = DefaultIncomingCallControls,
  landscape
}) => {
  const {
    t
  } = useI18n();
  const {
    theme: {
      colors,
      incomingCall,
      typefaces
    }
  } = useTheme();
  const landscapeContentStyles = {
    flexDirection: landscape ? 'row' : 'column'
  };
  return /*#__PURE__*/React.createElement(Background, null, /*#__PURE__*/React.createElement(View, {
    style: [styles.content, landscapeContentStyles, incomingCall.content]
  }, /*#__PURE__*/React.createElement(View, {
    style: [styles.topContainer, incomingCall.topContainer]
  }, /*#__PURE__*/React.createElement(UserInfo, null), /*#__PURE__*/React.createElement(Text, {
    style: [styles.incomingCallText, {
      color: colors.textPrimary
    }, typefaces.heading6, incomingCall.incomingCallText]
  }, t('Incoming Call...'))), /*#__PURE__*/React.createElement(View, {
    style: [styles.bottomContainer, incomingCall.bottomContainer]
  }, /*#__PURE__*/React.createElement(View, {
    style: [styles.incomingCallControls, incomingCall.incomingCallControls]
  }, IncomingCallControls && /*#__PURE__*/React.createElement(IncomingCallControls, {
    onAcceptCallHandler: onAcceptCallHandler,
    onRejectCallHandler: onRejectCallHandler
  })))));
};
const Background = ({
  children
}) => {
  const {
    theme: {
      colors,
      incomingCall
    }
  } = useTheme();
  const connectedUser = useConnectedUser();
  const {
    useCallMembers
  } = useCallStateHooks();
  const members = useCallMembers();

  // take the first N members to show their avatars
  const avatarsToShow = (members || []).filter(({
    user
  }) => user.id !== connectedUser?.id).map(({
    user
  }) => user.image).filter(image => !!image);
  if (avatarsToShow.length) {
    return /*#__PURE__*/React.createElement(ImageBackground, {
      source: {
        uri: avatarsToShow[0]
      },
      style: [styles.background, {
        backgroundColor: colors.sheetTertiary
      }, incomingCall.background]
    }, children);
  }
  return /*#__PURE__*/React.createElement(View, {
    style: [styles.background, {
      backgroundColor: colors.sheetTertiary
    }, incomingCall.background]
  }, children);
};
export const styles = StyleSheet.create({
  background: {
    flex: 1
  },
  content: {
    flex: 1
  },
  topContainer: {
    flex: 1,
    justifyContent: 'center'
  },
  incomingCallText: {
    marginTop: 8,
    textAlign: 'center'
  },
  bottomContainer: {
    flex: 1,
    justifyContent: 'center'
  },
  incomingCallControls: {
    justifyContent: 'center'
  }
});
