// @stream-io/video-react-native-sdk/dist/module/components/Call/RingingCallContent/OutgoingCall.js
import React from 'react';
import { StyleSheet, Text, View } from 'react-native';
import { UserInfo } from './UserInfo';
import { Z_INDEX } from '../../../constants';
import { useCallStateHooks, useI18n } from '@stream-io/video-react-bindings';
import { RTCView } from '@stream-io/react-native-webrtc';
import { useTheme } from '../../../contexts/ThemeContext';
import { OutgoingCallControls as DefaultOutgoingCallControls } from '../CallControls';

/**
 * Props for the OutgoingCall Component.
 */

/**
 * An outgoing call with the callee's avatar, name, caller's camera in background, reject and mute buttons.
 * Used after the user has initiated a call.
 */
export const OutgoingCall = ({
  OutgoingCallControls = DefaultOutgoingCallControls,
  landscape
}) => {
  const {
    theme: {
      colors,
      typefaces,
      outgoingCall
    }
  } = useTheme();
  const {
    t
  } = useI18n();
  const landscapeContentStyles = {
    flexDirection: landscape ? 'row' : 'column'
  };
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(View, {
    style: [StyleSheet.absoluteFill, styles.container, outgoingCall.container]
  }, /*#__PURE__*/React.createElement(View, {
    style: [styles.content, landscapeContentStyles, outgoingCall.content]
  }, /*#__PURE__*/React.createElement(View, {
    style: [styles.topContainer, outgoingCall.topContainer]
  }, /*#__PURE__*/React.createElement(UserInfo, null), /*#__PURE__*/React.createElement(Text, {
    style: [styles.callingText, {
      color: colors.textPrimary
    }, typefaces.heading6, outgoingCall.callingText]
  }, t('Calling...'))), /*#__PURE__*/React.createElement(View, {
    style: [styles.bottomContainer, outgoingCall.bottomContainer]
  }, /*#__PURE__*/React.createElement(View, {
    style: [styles.outgoingCallControls, outgoingCall.outgoingCallControls]
  }, OutgoingCallControls && /*#__PURE__*/React.createElement(OutgoingCallControls, null))))), /*#__PURE__*/React.createElement(Background, null));
};
const Background = () => {
  const {
    theme: {
      colors,
      outgoingCall
    }
  } = useTheme();
  const {
    useCameraState
  } = useCallStateHooks();
  const {
    isMute,
    camera
  } = useCameraState();
  const localVideoStream = camera.state.mediaStream;
  if (isMute || !localVideoStream) {
    return /*#__PURE__*/React.createElement(View, {
      style: [styles.background, {
        backgroundColor: colors.sheetSecondary
      }, outgoingCall.background]
    });
  }
  return /*#__PURE__*/React.createElement(View, {
    style: [styles.background, {
      backgroundColor: colors.sheetSecondary
    }, outgoingCall.background]
  }, /*#__PURE__*/React.createElement(RTCView, {
    streamURL: localVideoStream.toURL(),
    zOrder: Z_INDEX.IN_BACK,
    style: StyleSheet.absoluteFill,
    mirror: true,
    objectFit: "cover"
  }));
};
const styles = StyleSheet.create({
  background: {
    flex: 1
  },
  container: {
    zIndex: Z_INDEX.IN_MIDDLE
  },
  topContainer: {
    flex: 1,
    justifyContent: 'center'
  },
  content: {
    flex: 1
  },
  callingText: {
    marginTop: 16,
    textAlign: 'center'
  },
  bottomContainer: {
    flex: 1,
    alignSelf: 'center',
    justifyContent: 'center'
  },
  outgoingCallControls: {
    justifyContent: 'center'
  }
});
