// @stream-io/video-react-native-sdk/dist/module/components/Call/RingingCallContent/UserInfo.js
import React from 'react';
import { Image, StyleSheet, Text, View } from 'react-native';
import { generateCallTitle } from '../../../utils';
import { useCallStateHooks, useConnectedUser } from '@stream-io/video-react-bindings';
import { useTheme } from '../../../contexts/ThemeContext';
var AvatarModes = /*#__PURE__*/function (AvatarModes) {
  AvatarModes["small"] = "sm";
  AvatarModes["medium"] = "md";
  AvatarModes["large"] = "lg";
  return AvatarModes;
}(AvatarModes || {});
export const UserInfo = ({
  includeSelf = false,
  totalMembersToShow = 3
}) => {
  const {
    theme: {
      colors,
      typefaces,
      variants: {
        avatarSizes
      },
      userInfo
    }
  } = useTheme();
  const connectedUser = useConnectedUser();
  const {
    useCallMembers
  } = useCallStateHooks();
  const members = useCallMembers();

  // take the first N members to show their avatars
  const membersToShow = (members || []).filter(user => user.user_id !== connectedUser?.id || includeSelf).slice(0, totalMembersToShow).map(({
    user
  }) => user);
  if (includeSelf && !membersToShow.find(user => user.id === connectedUser?.id)) {
    // if the current user is not in the initial batch of members,
    // replace the first item in membersToShow array with the current user
    const self = members.find(({
      user
    }) => user.id === connectedUser?.id);
    if (self) {
      membersToShow.splice(0, 1, self.user);
    }
  }
  const memberUserIds = membersToShow.map(memberToShow => memberToShow.name ?? memberToShow.id);
  const callTitle = generateCallTitle(memberUserIds, totalMembersToShow);
  const avatarSizeModes = {
    1: AvatarModes.large,
    2: AvatarModes.medium,
    3: AvatarModes.small
  };
  const mode = avatarSizeModes[memberUserIds.length] || AvatarModes.small;
  const avatarStyles = {
    height: avatarSizes[mode],
    width: avatarSizes[mode],
    borderRadius: avatarSizes[mode] / 2,
    marginVertical: 4
  };
  const fontStyleByMembersCount = memberUserIds.length > 1 ? typefaces.heading5 : typefaces.heading4;
  return /*#__PURE__*/React.createElement(View, {
    style: [styles.container, userInfo.container]
  }, /*#__PURE__*/React.createElement(View, {
    style: [styles.avatarGroup, userInfo.avatarGroup]
  }, membersToShow.map(memberToShow => {
    if (!memberToShow.image) {
      return null;
    }
    return /*#__PURE__*/React.createElement(Image, {
      key: memberToShow.id,
      style: [avatarStyles]
      // FIXME: use real avatar from coordinator this is temporary
      ,
      source: {
        uri: memberToShow.image
      }
    });
  })), /*#__PURE__*/React.createElement(Text, {
    style: [styles.name, fontStyleByMembersCount, {
      color: colors.textPrimary
    }, userInfo.name]
  }, callTitle));
};
const styles = StyleSheet.create({
  container: {
    paddingHorizontal: 64,
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'space-between'
  },
  avatarGroup: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-evenly',
    flexWrap: 'wrap'
  },
  name: {
    textAlign: 'center',
    marginTop: 16
  }
});
