// @stream-io/video-react-native-sdk/dist/commonjs/components/Participant/ParticipantView/ParticipantNetworkQualityIndicator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ParticipantNetworkQualityIndicator = void 0;
var _reactNative = require("react-native");
var _reactNativeSvg = require("react-native-svg");
var _videoClient = require("@stream-io/video-client");
var _react = _interopRequireDefault(require("react"));
var _constants = require("../../../constants");
var _ThemeContext = require("../../../contexts/ThemeContext");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * Props for the NetworkQualityIndicator component.
 */

const useConnectionQualitySignalColors = participant => {
  const {
    theme: {
      colors
    }
  } = (0, _ThemeContext.useTheme)();
  const {
    connectionQuality
  } = participant;
  if (!connectionQuality) {
    return null;
  }
  switch (connectionQuality) {
    case _videoClient.SfuModels.ConnectionQuality.EXCELLENT:
      return [colors.iconSuccess, colors.iconSuccess, colors.iconSuccess];
    case _videoClient.SfuModels.ConnectionQuality.GOOD:
      return [colors.iconSuccess, colors.iconSuccess, colors.iconPrimary];
    case _videoClient.SfuModels.ConnectionQuality.POOR:
      return [colors.iconWarning, colors.iconPrimary, colors.iconPrimary];
    default:
      return null;
  }
};
const ParticipantNetworkQualityIndicator = ({
  participant
}) => {
  const {
    theme: {
      colors,
      variants: {
        iconSizes
      },
      participantNetworkQualityIndicator
    }
  } = (0, _ThemeContext.useTheme)();
  const connectionQualityColors = useConnectionQualitySignalColors(participant);
  if (!connectionQualityColors) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, {
      backgroundColor: colors.sheetOverlay,
      height: iconSizes.lg,
      width: iconSizes.lg
    }, participantNetworkQualityIndicator.container]
  }, /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Svg, {
    viewBox: "0 0 34 34",
    fill: 'none'
  }, /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Path, {
    d: "M 9.97559 22.3379 L 9.97559 19.616",
    stroke: connectionQualityColors[0],
    strokeWidth: 3,
    strokeLinecap: "round",
    strokeLinejoin: "round",
    fill: connectionQualityColors[0]
  }), /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Path, {
    d: "M 16.7808 22.3379 L 16.7808 15.5331",
    stroke: connectionQualityColors[1],
    strokeWidth: 3,
    strokeLinecap: "round",
    strokeLinejoin: "round",
    fill: connectionQualityColors[1]
  }), /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Path, {
    d: "M 23.5854 22.3379 L 23.5854 11.4502",
    stroke: connectionQualityColors[2],
    strokeWidth: 3,
    strokeLinecap: "round",
    strokeLinejoin: "round",
    fill: connectionQualityColors[2]
  })));
};
exports.ParticipantNetworkQualityIndicator = ParticipantNetworkQualityIndicator;
const styles = _reactNative.StyleSheet.create({
  container: {
    zIndex: _constants.Z_INDEX.IN_FRONT,
    alignSelf: 'flex-end',
    borderTopLeftRadius: 5
  }
});
