// @stream-io/video-react-native-sdk/dist/commonjs/theme/colors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.colors = exports.colorPalette = void 0;
var _constants = require("./constants");
const colors = exports.colors = {
  primary: _constants.palette.primary100,
  secondary: _constants.palette.neutral90,
  success: _constants.palette.success100,
  warning: _constants.palette.warning100,
  buttonPrimary: _constants.palette.primary100,
  buttonSecondary: _constants.palette.neutral90,
  buttonSuccess: _constants.palette.success100,
  buttonWarning: _constants.palette.warning100,
  buttonDisabled: _constants.palette.primary180,
  iconPrimary: _constants.palette.neutral0,
  iconSecondary: _constants.palette.primary100,
  iconSuccess: _constants.palette.success100,
  iconWarning: _constants.palette.warning100,
  sheetPrimary: _constants.palette.neutral120,
  sheetSecondary: _constants.palette.neutral110,
  sheetTertiary: _constants.palette.neutral90,
  sheetOverlay: _constants.palette.overlay,
  textPrimary: _constants.palette.neutral0,
  textSecondary: _constants.palette.neutral30
};
const colorPalette = exports.colorPalette = {
  colors
};
