// @stream-io/video-react-native-sdk/dist/commonjs/components/Call/CallLayout/CallParticipantsGrid.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CallParticipantsGrid = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _videoReactBindings = require("@stream-io/video-react-bindings");
var _useDebouncedValue = require("../../../utils/hooks/useDebouncedValue");
var _CallParticipantsList = require("../CallParticipantsList/CallParticipantsList");
var _TestIds = require("../../../constants/TestIds");
var _ThemeContext = require("../../../contexts/ThemeContext");
var _useIsInPiPMode = require("../../../hooks/useIsInPiPMode");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
/**
 * Props for the CallParticipantsGrid component.
 */

/**
 * Component used to display the list of participants in a grid mode.
 */
const CallParticipantsGrid = ({
  CallParticipantsList = _CallParticipantsList.CallParticipantsList,
  ParticipantLabel,
  ParticipantNetworkQualityIndicator,
  ParticipantReaction,
  ParticipantVideoFallback,
  ParticipantView,
  VideoRenderer,
  showLocalParticipant = false,
  supportedReactions,
  landscape
}) => {
  const {
    theme: {
      colors,
      callParticipantsGrid
    }
  } = (0, _ThemeContext.useTheme)();
  const {
    useRemoteParticipants,
    useParticipants,
    useLocalParticipant,
    useDominantSpeaker
  } = (0, _videoReactBindings.useCallStateHooks)();
  const _remoteParticipants = useRemoteParticipants();
  const localParticipant = useLocalParticipant();
  const _allParticipants = useParticipants();
  const dominantSpeaker = useDominantSpeaker();
  // we debounce the participants arrays to avoid unnecessary rerenders that happen when participant tracks are all subscribed simultaneously
  const remoteParticipants = (0, _useDebouncedValue.useDebouncedValue)(_remoteParticipants, 300);
  const allParticipants = (0, _useDebouncedValue.useDebouncedValue)(_allParticipants, 300);
  const landscapeStyles = {
    flexDirection: landscape ? 'row' : 'column'
  };
  const isInPiPMode = (0, _useIsInPiPMode.useIsInPiPMode)();
  const showFloatingView = !isInPiPMode && remoteParticipants.length > 0 && remoteParticipants.length < 3;
  let participants;
  if (showFloatingView) {
    if (showLocalParticipant && localParticipant) {
      participants = [localParticipant];
    } else {
      participants = remoteParticipants;
    }
  } else {
    participants = allParticipants;
  }
  if (isInPiPMode) {
    if (dominantSpeaker && !dominantSpeaker.isLocalParticipant) {
      participants = [dominantSpeaker];
    } else if (remoteParticipants[0]) {
      participants = [remoteParticipants[0]];
    } else if (localParticipant) {
      participants = [localParticipant];
    } else {
      participants = [];
    }
  }
  const participantViewProps = {
    ParticipantView,
    ParticipantLabel,
    ParticipantNetworkQualityIndicator,
    ParticipantReaction,
    ParticipantVideoFallback,
    VideoRenderer
  };
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, landscapeStyles, {
      backgroundColor: colors.sheetPrimary
    }, callParticipantsGrid.container],
    testID: _TestIds.ComponentTestIds.CALL_PARTICIPANTS_GRID
  }, CallParticipantsList && /*#__PURE__*/_react.default.createElement(CallParticipantsList, _extends({
    participants: participants,
    supportedReactions: supportedReactions,
    landscape: landscape
  }, participantViewProps)));
};
exports.CallParticipantsGrid = CallParticipantsGrid;
const styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1
  }
});
