// @stream-io/video-react-native-sdk/dist/module/components/Call/CallControls/CallControls.js
import React from 'react';
import { StyleSheet, View } from 'react-native';
import { ToggleAudioPublishingButton } from './ToggleAudioPublishingButton';
import { ToggleVideoPublishingButton } from './ToggleVideoPublishingButton';
import { ToggleCameraFaceButton } from './ToggleCameraFaceButton';
import { Z_INDEX } from '../../../constants';
import { HangUpCallButton } from './HangupCallButton';
import { useTheme } from '../../../contexts/ThemeContext';

/**
 * Props for the CallControls Component.
 */

/**
 * A list/row of controls (mute audio/video, toggle front/back camera, hangup call etc.)
 * the user can trigger within an active call.
 */
export const CallControls = ({
  style,
  onHangupCallHandler,
  landscape
}) => {
  const {
    theme: {
      colors,
      callControls
    }
  } = useTheme();
  const landscapeStyles = {
    flexDirection: 'row',
    paddingHorizontal: landscape ? 12 : 0,
    paddingVertical: landscape ? 0 : 12
  };
  return /*#__PURE__*/React.createElement(View, {
    style: [styles.container, {
      backgroundColor: colors.sheetPrimary
    }, callControls.container, landscapeStyles, style]
  }, /*#__PURE__*/React.createElement(ToggleVideoPublishingButton, null), /*#__PURE__*/React.createElement(ToggleAudioPublishingButton, null), /*#__PURE__*/React.createElement(ToggleCameraFaceButton, null), /*#__PURE__*/React.createElement(HangUpCallButton, {
    onPressHandler: onHangupCallHandler
  }));
};
const styles = StyleSheet.create({
  container: {
    justifyContent: 'space-evenly',
    zIndex: Z_INDEX.IN_FRONT
  }
});
