// @stream-io/video-react-native-sdk/dist/module/components/Call/CallControls/IncomingCallControls.js
import React from 'react';
import { StyleSheet, View } from 'react-native';
import { useTheme } from '../../../contexts';
import { AcceptCallButton } from './AcceptCallButton';
import { RejectCallButton } from './RejectCallButton';

/**
 * Props for the IncomingCallControls Component.
 */

export const IncomingCallControls = ({
  onAcceptCallHandler,
  onRejectCallHandler
}) => {
  const {
    theme: {
      incomingCall,
      variants: {
        buttonSizes
      }
    }
  } = useTheme();
  return /*#__PURE__*/React.createElement(View, {
    style: [styles.buttonGroup, incomingCall.buttonGroup]
  }, /*#__PURE__*/React.createElement(RejectCallButton, {
    onPressHandler: onRejectCallHandler,
    size: buttonSizes.md,
    rejectReason: "decline"
  }), /*#__PURE__*/React.createElement(AcceptCallButton, {
    onPressHandler: onAcceptCallHandler
  }));
};
const styles = StyleSheet.create({
  buttonGroup: {
    flexDirection: 'row',
    justifyContent: 'space-evenly',
    alignItems: 'center'
  }
});
