// @stream-io/video-react-native-sdk/dist/module/components/Call/CallControls/OutgoingCallControls.js
import React from 'react';
import { StyleSheet, View } from 'react-native';
import { useTheme } from '../../../contexts';
import { ToggleAudioPreviewButton } from './ToggleAudioPreviewButton';
import { ToggleVideoPreviewButton } from './ToggleVideoPreviewButton';
import { RejectCallButton } from './RejectCallButton';

/**
 * Props for the OutgoingCallControls Component.
 */

export const OutgoingCallControls = ({
  onHangupCallHandler
}) => {
  const {
    theme: {
      outgoingCall,
      variants: {
        buttonSizes
      }
    }
  } = useTheme();
  return /*#__PURE__*/React.createElement(View, {
    style: [styles.buttonGroup, outgoingCall.buttonGroup]
  }, /*#__PURE__*/React.createElement(View, {
    style: [styles.deviceControlButtons, outgoingCall.deviceControlButtons]
  }, /*#__PURE__*/React.createElement(ToggleAudioPreviewButton, null), /*#__PURE__*/React.createElement(ToggleVideoPreviewButton, null)), /*#__PURE__*/React.createElement(RejectCallButton, {
    onPressHandler: onHangupCallHandler,
    size: buttonSizes.md,
    rejectReason: "cancel"
  }));
};
const styles = StyleSheet.create({
  buttonGroup: {
    alignItems: 'center'
  },
  deviceControlButtons: {
    width: '100%',
    flexDirection: 'row',
    justifyContent: 'space-evenly'
  }
});
