// @stream-io/video-react-native-sdk/dist/module/components/Call/CallControls/RejectCallButton.js
import { useCall, useCallStateHooks } from '@stream-io/video-react-bindings';
import React, { useState } from 'react';
import { CallControlsButton } from './CallControlsButton';
import { IconWrapper, PhoneDown } from '../../../icons';
import { CallingState, getLogger } from '@stream-io/video-client';
import { useTheme } from '../../../contexts/ThemeContext';

/**
 * The props for the Reject Call button.
 */

/**
 * Button to reject a call.
 *
 * Calls call.leave({ reject: true, reason: `OPTIONAL-REASON` }) internally.
 */
export const RejectCallButton = ({
  onPressHandler,
  onRejectCallHandler,
  size,
  rejectReason
}) => {
  const call = useCall();
  const {
    useCallCallingState
  } = useCallStateHooks();
  const callingState = useCallCallingState();
  const {
    theme: {
      colors,
      rejectCallButton,
      variants: {
        buttonSizes,
        iconSizes
      }
    }
  } = useTheme();
  const [isLoading, setIsLoading] = useState(false);
  const rejectCallHandler = async () => {
    setIsLoading(true);
    if (onPressHandler) {
      onPressHandler();
      return;
    }
    try {
      if (callingState === CallingState.LEFT) {
        return;
      }
      await call?.leave({
        reject: true,
        reason: rejectReason
      });
      if (onRejectCallHandler) {
        onRejectCallHandler();
      }
    } catch (error) {
      const logger = getLogger(['RejectCallButton']);
      logger('error', 'Error rejecting Call', error);
    } finally {
      setIsLoading(false);
    }
  };
  return /*#__PURE__*/React.createElement(CallControlsButton, {
    onPress: rejectCallHandler,
    color: colors.buttonWarning,
    size: size ?? buttonSizes.md
    // TODO: check what to do about this random style prop
    // svgContainerStyle={theme.icon.lg}
    ,
    style: rejectCallButton,
    disabled: isLoading
  }, /*#__PURE__*/React.createElement(IconWrapper, null, /*#__PURE__*/React.createElement(PhoneDown, {
    color: colors.iconPrimary,
    size: iconSizes.lg
  })));
};
