// @stream-io/video-react-native-sdk/dist/module/components/Call/CallControls/ToggleVideoPreviewButton.js
import React from 'react';
import { useCallStateHooks } from '@stream-io/video-react-bindings';
import { useTheme } from '../../../contexts';
import { CallControlsButton } from './CallControlsButton';
import { IconWrapper, Video, VideoSlash } from '../../../icons';

/**
 * Props for the Toggle Video preview button
 */

/**
 * Button to toggle video mute/unmute status before joining the call.
 */
export const ToggleVideoPreviewButton = ({
  onPressHandler
}) => {
  const {
    theme: {
      colors,
      toggleVideoPreviewButton,
      variants: {
        buttonSizes,
        iconSizes
      }
    }
  } = useTheme();
  const {
    useCameraState,
    useCallSettings
  } = useCallStateHooks();
  const callSettings = useCallSettings();
  const isVideoEnabledInCall = callSettings?.video.enabled;
  const {
    optimisticIsMute,
    camera
  } = useCameraState();
  const onPress = async () => {
    if (onPressHandler) {
      onPressHandler();
      return;
    }
    await camera.toggle();
  };
  if (!isVideoEnabledInCall) {
    return;
  }
  return /*#__PURE__*/React.createElement(CallControlsButton, {
    onPress: onPress,
    color: !optimisticIsMute ? colors.buttonSecondary : colors.buttonWarning,
    size: buttonSizes.md,
    style: {
      container: toggleVideoPreviewButton.container,
      svgContainer: toggleVideoPreviewButton.svgContainer
    }
  }, /*#__PURE__*/React.createElement(IconWrapper, null, !optimisticIsMute ? /*#__PURE__*/React.createElement(Video, {
    color: colors.iconPrimary,
    size: iconSizes.lg
  }) : /*#__PURE__*/React.createElement(VideoSlash, {
    color: colors.iconPrimary,
    size: iconSizes.lg
  })));
};
