// @stream-io/video-react-native-sdk/dist/module/components/Call/CallControls/ToggleAudioPreviewButton.js
import { useCallStateHooks } from '@stream-io/video-react-bindings';
import React from 'react';
import { useTheme } from '../../../contexts';
import { IconWrapper, Mic, MicOff } from '../../../icons';
import { CallControlsButton } from './CallControlsButton';

/**
 * Props for the Toggle Audio preview button
 */

/**
 * Button to toggle audio mute/unmute status before joining the call.
 */
export const ToggleAudioPreviewButton = ({
  onPressHandler
}) => {
  const {
    theme: {
      colors,
      toggleAudioPreviewButton,
      variants: {
        buttonSizes
      },
      defaults
    }
  } = useTheme();
  const {
    useMicrophoneState
  } = useCallStateHooks();
  const {
    optimisticIsMute,
    microphone
  } = useMicrophoneState();
  const onPress = async () => {
    if (onPressHandler) {
      onPressHandler();
      return;
    }
    await microphone.toggle();
  };
  return /*#__PURE__*/React.createElement(CallControlsButton, {
    onPress: onPress,
    color: !optimisticIsMute ? colors.buttonSecondary : colors.buttonWarning,
    size: buttonSizes.md,
    style: {
      container: toggleAudioPreviewButton.container,
      svgContainer: toggleAudioPreviewButton.svgContainer
    }
  }, /*#__PURE__*/React.createElement(IconWrapper, null, !optimisticIsMute ? /*#__PURE__*/React.createElement(Mic, {
    color: colors.iconPrimary,
    size: defaults.iconSize
  }) : /*#__PURE__*/React.createElement(MicOff, {
    color: colors.iconPrimary,
    size: defaults.iconSize
  })));
};
