// @stream-io/video-react-native-sdk/dist/module/components/Livestream/LivestreamTopView/HostLivestreamTopView.js
import React from 'react';
import { StyleSheet, View } from 'react-native';
import { DurationBadge as DefaultDurationBadge } from './DurationBadge';
import { LiveIndicator as DefaultLiveIndicator } from './LiveIndicator';
import { FollowerCount as DefaultFollowerCount } from './FollowerCount';
import { useTheme } from '../../../contexts';
import { useCallStateHooks } from '@stream-io/video-react-bindings';
import { Z_INDEX } from '../../../constants';

/**
 * Props for the HostLivestreamTopView component.
 */

/**
 * The HostLivestreamTopView component displays the top view component of the host's live stream.
 */
export const HostLivestreamTopView = ({
  DurationBadge = DefaultDurationBadge,
  LiveIndicator = DefaultLiveIndicator,
  FollowerCount = DefaultFollowerCount,
  onLayout
}) => {
  const {
    useIsCallLive,
    useIsCallHLSBroadcastingInProgress
  } = useCallStateHooks();
  const isCallLive = useIsCallLive();
  const isBroadcasting = useIsCallHLSBroadcastingInProgress();
  const liveOrBroadcasting = isCallLive || isBroadcasting;
  const {
    theme: {
      colors,
      hostLivestreamTopView
    }
  } = useTheme();
  return /*#__PURE__*/React.createElement(View, {
    style: [styles.container, {
      backgroundColor: colors.sheetOverlay
    }, hostLivestreamTopView.container],
    onLayout: onLayout
  }, /*#__PURE__*/React.createElement(View, {
    style: [styles.leftElement, hostLivestreamTopView.leftElement]
  }, DurationBadge && /*#__PURE__*/React.createElement(DurationBadge, {
    mode: "host"
  })), /*#__PURE__*/React.createElement(View, {
    style: [styles.centerElement, hostLivestreamTopView.centerElement]
  }), /*#__PURE__*/React.createElement(View, {
    style: [styles.rightElement, hostLivestreamTopView.rightElement]
  }, /*#__PURE__*/React.createElement(View, {
    style: [styles.liveInfo, hostLivestreamTopView.liveInfo]
  }, liveOrBroadcasting && LiveIndicator && /*#__PURE__*/React.createElement(LiveIndicator, null), FollowerCount && /*#__PURE__*/React.createElement(FollowerCount, null))));
};
const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    top: 0,
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: 16,
    paddingHorizontal: 8,
    borderBottomEndRadius: 8,
    borderBottomStartRadius: 8,
    zIndex: Z_INDEX.IN_FRONT
  },
  liveInfo: {
    flexDirection: 'row'
  },
  leftElement: {
    flex: 1,
    alignItems: 'flex-start'
  },
  centerElement: {
    flex: 1,
    alignItems: 'center'
  },
  rightElement: {
    flex: 1,
    alignItems: 'flex-end'
  }
});
