// @stream-io/video-react-native-sdk/dist/module/components/Livestream/LivestreamTopView/ViewerLivestreamTopView.js
import React from 'react';
import { StyleSheet, View } from 'react-native';
import { DurationBadge as DefaultDurationBadge } from './DurationBadge';
import { LiveIndicator as DefaultLiveIndicator } from './LiveIndicator';
import { FollowerCount as DefaultFollowerCount } from './FollowerCount';
import { useTheme } from '../../../contexts';
import { Z_INDEX } from '../../../constants';

/**
 * Props for the ViewerLivestreamTopView component.
 */

/**
 * The ViewerLivestreamTopView component displays the top view component of the viewer's live stream.
 */
export const ViewerLivestreamTopView = ({
  DurationBadge = DefaultDurationBadge,
  LiveIndicator = DefaultLiveIndicator,
  FollowerCount = DefaultFollowerCount,
  onLayout
}) => {
  const {
    theme: {
      colors,
      viewerLivestreamTopView
    }
  } = useTheme();
  return /*#__PURE__*/React.createElement(View, {
    style: [styles.container, {
      backgroundColor: colors.sheetOverlay
    }, viewerLivestreamTopView.container],
    onLayout: onLayout
  }, /*#__PURE__*/React.createElement(View, {
    style: [styles.leftElement, viewerLivestreamTopView.leftElement]
  }, /*#__PURE__*/React.createElement(View, {
    style: [styles.liveInfo, viewerLivestreamTopView.liveInfo]
  }, LiveIndicator && /*#__PURE__*/React.createElement(LiveIndicator, null), FollowerCount && /*#__PURE__*/React.createElement(FollowerCount, null))), /*#__PURE__*/React.createElement(View, {
    style: [styles.centerElement, viewerLivestreamTopView.centerElement]
  }, DurationBadge && /*#__PURE__*/React.createElement(DurationBadge, {
    mode: "viewer"
  })), /*#__PURE__*/React.createElement(View, {
    style: [styles.rightElement, viewerLivestreamTopView.rightElement]
  }));
};
const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    top: 0,
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: 16,
    paddingHorizontal: 8,
    borderBottomEndRadius: 8,
    borderBottomStartRadius: 8,
    zIndex: Z_INDEX.IN_FRONT
  },
  liveInfo: {
    flexDirection: 'row'
  },
  leftElement: {
    flex: 1,
    alignItems: 'flex-start'
  },
  centerElement: {
    flex: 1,
    alignItems: 'center',
    flexGrow: 3
  },
  rightElement: {
    flex: 1,
    alignItems: 'flex-end'
  }
});
