// @stream-io/video-react-native-sdk/dist/commonjs/components/Call/CallControls/IncomingCallControls.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.IncomingCallControls = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _contexts = require("../../../contexts");
var _AcceptCallButton = require("./AcceptCallButton");
var _RejectCallButton = require("./RejectCallButton");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * Props for the IncomingCallControls Component.
 */

const IncomingCallControls = ({
  onAcceptCallHandler,
  onRejectCallHandler
}) => {
  const {
    theme: {
      incomingCall,
      variants: {
        buttonSizes
      }
    }
  } = (0, _contexts.useTheme)();
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.buttonGroup, incomingCall.buttonGroup]
  }, /*#__PURE__*/_react.default.createElement(_RejectCallButton.RejectCallButton, {
    onPressHandler: onRejectCallHandler,
    size: buttonSizes.md,
    rejectReason: "decline"
  }), /*#__PURE__*/_react.default.createElement(_AcceptCallButton.AcceptCallButton, {
    onPressHandler: onAcceptCallHandler
  }));
};
exports.IncomingCallControls = IncomingCallControls;
const styles = _reactNative.StyleSheet.create({
  buttonGroup: {
    flexDirection: 'row',
    justifyContent: 'space-evenly',
    alignItems: 'center'
  }
});
