// @stream-io/video-react-native-sdk/dist/commonjs/components/Call/CallControls/LobbyControls.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LobbyControls = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _ToggleAudioPreviewButton = require("./ToggleAudioPreviewButton");
var _ToggleVideoPreviewButton = require("./ToggleVideoPreviewButton");
var _ThemeContext = require("../../../contexts/ThemeContext");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
/**
 * Controls for the Lobby Component
 */
const LobbyControls = () => {
  const {
    theme: {
      lobbyControls
    }
  } = (0, _ThemeContext.useTheme)();
  const styles = useStyles();
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, lobbyControls.container]
  }, /*#__PURE__*/_react.default.createElement(_ToggleAudioPreviewButton.ToggleAudioPreviewButton, null), /*#__PURE__*/_react.default.createElement(_ToggleVideoPreviewButton.ToggleVideoPreviewButton, null));
};
exports.LobbyControls = LobbyControls;
const useStyles = () => {
  const {
    theme
  } = (0, _ThemeContext.useTheme)();
  return (0, _react.useMemo)(() => _reactNative.StyleSheet.create({
    container: {
      paddingTop: theme.variants.spacingSizes.xs,
      flexDirection: 'row',
      justifyContent: 'space-evenly'
    }
  }), [theme]);
};
