// @stream-io/video-react-native-sdk/dist/commonjs/components/Call/CallControls/ToggleVideoPublishingButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ToggleVideoPublishingButton = void 0;
var _react = _interopRequireDefault(require("react"));
var _videoClient = require("@stream-io/video-client");
var _videoReactBindings = require("@stream-io/video-react-bindings");
var _CallControlsButton = require("./CallControlsButton");
var _icons = require("../../../icons");
var _ThemeContext = require("../../../contexts/ThemeContext");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * Props for the Toggle Video publishing button
 */

/**
 * Button to toggle video mute/unmute status while in the call.
 */
const ToggleVideoPublishingButton = ({
  onPressHandler
}) => {
  const {
    useCameraState,
    useCallSettings
  } = (0, _videoReactBindings.useCallStateHooks)();
  const {
    camera,
    optimisticIsMute
  } = useCameraState();
  const callSettings = useCallSettings();
  const isVideoEnabledInCall = callSettings?.video.enabled;
  const {
    theme: {
      colors,
      variants
    }
  } = (0, _ThemeContext.useTheme)();
  const onPress = async () => {
    if (onPressHandler) {
      onPressHandler();
      return;
    }
    await camera.toggle();
  };
  if (!isVideoEnabledInCall) {
    return;
  }
  return /*#__PURE__*/_react.default.createElement(_videoReactBindings.Restricted, {
    requiredGrants: [_videoClient.OwnCapability.SEND_VIDEO]
  }, /*#__PURE__*/_react.default.createElement(_CallControlsButton.CallControlsButton, {
    onPress: onPress,
    color: !optimisticIsMute ? colors.buttonSecondary : colors.buttonWarning
  }, /*#__PURE__*/_react.default.createElement(_icons.IconWrapper, null, !optimisticIsMute ? /*#__PURE__*/_react.default.createElement(_icons.Video, {
    color: colors.iconPrimary,
    size: variants.iconSizes.md
  }) : /*#__PURE__*/_react.default.createElement(_icons.VideoSlash, {
    color: colors.iconPrimary,
    size: variants.iconSizes.md
  }))));
};
exports.ToggleVideoPublishingButton = ToggleVideoPublishingButton;
