// @stream-io/video-react-native-sdk/dist/commonjs/components/Call/CallControls/ToggleAudioPreviewButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ToggleAudioPreviewButton = void 0;
var _videoReactBindings = require("@stream-io/video-react-bindings");
var _react = _interopRequireDefault(require("react"));
var _contexts = require("../../../contexts");
var _icons = require("../../../icons");
var _CallControlsButton = require("./CallControlsButton");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * Props for the Toggle Audio preview button
 */

/**
 * Button to toggle audio mute/unmute status before joining the call.
 */
const ToggleAudioPreviewButton = ({
  onPressHandler
}) => {
  const {
    theme: {
      colors,
      toggleAudioPreviewButton,
      variants: {
        buttonSizes
      },
      defaults
    }
  } = (0, _contexts.useTheme)();
  const {
    useMicrophoneState
  } = (0, _videoReactBindings.useCallStateHooks)();
  const {
    optimisticIsMute,
    microphone
  } = useMicrophoneState();
  const onPress = async () => {
    if (onPressHandler) {
      onPressHandler();
      return;
    }
    await microphone.toggle();
  };
  return /*#__PURE__*/_react.default.createElement(_CallControlsButton.CallControlsButton, {
    onPress: onPress,
    color: !optimisticIsMute ? colors.buttonSecondary : colors.buttonWarning,
    size: buttonSizes.md,
    style: {
      container: toggleAudioPreviewButton.container,
      svgContainer: toggleAudioPreviewButton.svgContainer
    }
  }, /*#__PURE__*/_react.default.createElement(_icons.IconWrapper, null, !optimisticIsMute ? /*#__PURE__*/_react.default.createElement(_icons.Mic, {
    color: colors.iconPrimary,
    size: defaults.iconSize
  }) : /*#__PURE__*/_react.default.createElement(_icons.MicOff, {
    color: colors.iconPrimary,
    size: defaults.iconSize
  })));
};
exports.ToggleAudioPreviewButton = ToggleAudioPreviewButton;
