// @stream-io/video-react-native-sdk/dist/module/components/utility/Avatar.js
import { Image, StyleSheet, Text, View } from 'react-native';
import React from 'react';
import { getInitialsOfName } from '../../utils';
import { ComponentTestIds, ImageTestIds } from '../../constants/TestIds';
import { useTheme } from '../../contexts/ThemeContext';

/**
 * Props to be passed for the Avatar component.
 */

/**
 * Shows either user's image or initials based on the user state and existence of
 * their image.
 */
export const Avatar = props => {
  const {
    participant: {
      userId,
      image,
      name
    },
    size = 100,
    style: styleProp
  } = props;
  const {
    theme: {
      avatar,
      colors,
      typefaces
    }
  } = useTheme();
  const userDetails = name || userId;
  const userLabel = userDetails ? getInitialsOfName(userDetails) : '?';
  const imageUrl = image;
  return /*#__PURE__*/React.createElement(View, {
    testID: ComponentTestIds.PARTICIPANT_AVATAR,
    style: [styles.container, {
      borderRadius: size / 2,
      height: size,
      width: size
    }, {
      backgroundColor: colors.primary
    }, avatar.container, styleProp?.container]
  }, imageUrl ? /*#__PURE__*/React.createElement(Image, {
    testID: ImageTestIds.AVATAR,
    source: {
      uri: imageUrl
    },
    style: [styles.image, avatar.image, styleProp?.image]
  }) : /*#__PURE__*/React.createElement(Text, {
    style: [styles.text, {
      fontSize: size / 2,
      color: colors.textPrimary
    }, typefaces.heading6, avatar.text, styleProp?.text],
    numberOfLines: 1
  }, userLabel));
};
const styles = StyleSheet.create({
  container: {
    justifyContent: 'center',
    overflow: 'hidden'
  },
  image: {
    flex: 1
  },
  text: {
    textAlign: 'center'
  }
});
